File: extensions/music/src/main/java/app/morphe/extension/music/patches/BypassImageRegionRestrictionsPatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * Original hard forked code:
 * https://github.com/inotia00/revanced-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to Morphe contributions.
 */

package app.morphe.extension.music.patches;

import java.util.regex.Pattern;

import app.morphe.extension.shared.Logger;

/**
 * Always-on patch.  Rewrites known YouTube static image hosts to a region-agnostic
 * domain so user avatars and channel images load in regions where the original host
 * is blocked.
 */
@SuppressWarnings("unused")
public final class BypassImageRegionRestrictionsPatch {

    private static final String REPLACEMENT_IMAGE_DOMAIN = "https://yt4.ggpht.com";

    /**
     * YouTube static images' domain.  Includes user and channel avatar images and community
     * post images.  Pattern mirrors RVX's broader matcher to cover all host shards.
     */
    private static final Pattern YOUTUBE_STATIC_IMAGE_DOMAIN_PATTERN = Pattern.compile(
            "^https://(ap[1-2]|gm[1-4]|gz0|(cp|ci|gp|lh)[3-6]|sp[1-3]|yt[3-4]|(play|ccp)-lh)\\.(ggpht|googleusercontent)\\.com"
    );

    /**
     * Injection point.  Called off the main thread and by multiple threads at the same time.
     *
     * @param originalUrl Image URL for all image URLs loaded.
     */
    public static String overrideImageURL(String originalUrl) {
        try {
            final String replacement = YOUTUBE_STATIC_IMAGE_DOMAIN_PATTERN
                    .matcher(originalUrl).replaceFirst(REPLACEMENT_IMAGE_DOMAIN);
            if (!replacement.equals(originalUrl)) {
                Logger.printDebug(() -> "Replaced: '" + originalUrl + "' with: '" + replacement + "'");
            }
            return replacement;
        } catch (Exception ex) {
            Logger.printException(() -> "overrideImageURL failure", ex);
        }
        return originalUrl;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/music/misc/imageregion/BypassImageRegionRestrictionsPatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * Original hard forked code:
 * https://github.com/inotia00/revanced-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to Morphe contributions.
 */

package app.morphe.patches.music.misc.imageregion

import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.music.misc.extension.sharedExtensionPatch
import app.morphe.patches.music.misc.imageurlhook.addImageURLHook
import app.morphe.patches.music.misc.imageurlhook.cronetImageURLHookPatch
import app.morphe.patches.music.shared.Constants.COMPATIBILITY_YOUTUBE_MUSIC_9X

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/music/patches/BypassImageRegionRestrictionsPatch;"

@Suppress("unused")
val bypassImageRegionRestrictionsPatch = bytecodePatch(
    name = "Bypass image region restrictions",
    description = "Uses a different host for user avatar and channel images so they load correctly " +
        "in regions where the default host is blocked.",
) {
    dependsOn(
        sharedExtensionPatch,
        cronetImageURLHookPatch,
    )

    compatibleWith(COMPATIBILITY_YOUTUBE_MUSIC_9X)

    execute {
        // Always-on: no setting check.  A priority hook is not needed, as the image URLs of interest
        // are not modified by any other patch in this repo.
        addImageURLHook(EXTENSION_CLASS)
    }
}
